File: experiments/issue73_20260908/counterword_search.py
#!/usr/bin/env python3
"""H-20260908-03: exhaustive U=A search on frozen unused periods.

A positive-sum word with every addition P2-supplied (unrestricted lag
1..p(p+1)) refutes E-067 and E-070. Clean exhaustive ranges are COMPUTED.
"""
from __future__ import annotations

import argparse
import hashlib
import json
import subprocess
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]


def all_supplied(word: tuple[int, ...]) -> tuple[bool, dict[int, list[int]]]:
    p = len(word)
    s = sum(word)
    if s <= 0:
        return False, {}
    # prefix on two periods for short lags, then closed form for wraps
    sig: dict[int, list[int]] = {}
    for t, step in enumerate(word):
        if step < 0:
            continue
        back = [word[(t - i) % p] for i in range(1, p + 1)]
        weight = sum(i * e for i, e in enumerate(back, 1))
        prefix = moment = 0
        hits: list[int] = []
        for r in range(p):
            if s != 0:
                quot, rem = divmod(1 - prefix, s)
                d = quot * p + r
                if rem == 0 and quot >= 0 and d > 0:
                    value = (
                        quot * weight
                        + p * s * quot * (quot - 1) // 2
                        + quot * p * prefix
                        + moment
                    )
                    if value == 0:
                        hits.append(d)
            prefix += back[r]
            moment += (r + 1) * back[r]
        if not hits:
            return False, {}
        sig[t] = hits
    return True, sig


def naive_hits(word: tuple[int, ...], t: int) -> list[int]:
    p = len(word)
    total = moment = 0
    hits = []
    for d in range(1, p * (p + 1) + 1):
        e = word[(t - d) % p]
        total += e
        moment += d * e
        if total == 1 and moment == 0:
            hits.append(d)
    return hits


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--min-period", type=int, required=True)
    parser.add_argument("--max-period", type=int, required=True)
    parser.add_argument("--output", type=Path, required=True)
    args = parser.parse_args()
    args.output.parent.mkdir(parents=True, exist_ok=True)
    print("protocol=H-20260908-03 exhaustive U=A counterword search", flush=True)
    print(
        "source_revision="
        + subprocess.check_output(["git", "rev-parse", "HEAD"], text=True).strip(),
        flush=True,
    )
    print(
        "script_sha256="
        + hashlib.sha256(Path(__file__).read_bytes()).hexdigest(),
        flush=True,
    )
    print(f"period_range=({args.min_period},{args.max_period})", flush=True)

    # Negative control: SSSSAAAASAAA must not be reported as U=A.
    control = tuple(1 if c == "A" else -1 for c in "SSSSAAAASAAA")
    ok, sig = all_supplied(control)
    print("control_SSSSAAAASAAA_UA=" + json.dumps(ok), flush=True)
    if ok:
        raise SystemExit("control failed: known unsupplied word reported as U=A")

    total_pos = 0
    ua = 0
    missing_hist = {}
    for p in range(args.min_period, args.max_period + 1):
        npos = 0
        nua = 0
        limit = 1 << p
        for mask in range(limit):
            word = tuple(1 if (mask >> i) & 1 else -1 for i in range(p))
            if sum(word) <= 0:
                continue
            npos += 1
            supplied, sig = all_supplied(word)
            if supplied:
                nua += 1
                # independent naive replay
                replay = {t: naive_hits(word, t) for t, e in enumerate(word) if e > 0}
                assert all(replay[t] for t in replay)
                witness = {
                    "period": p,
                    "word": "".join("A" if e > 0 else "S" for e in word),
                    "sign_sum": sum(word),
                    "mask": mask,
                    "suppliers": sig,
                    "naive": replay,
                    "positive_words_before": total_pos + npos,
                }
                args.output.write_text(json.dumps(witness, indent=2, sort_keys=True) + "\n")
                print("COUNTEREXAMPLE=" + json.dumps(witness, sort_keys=True), flush=True)
                print("STOP: E-067/E-070 refuted", flush=True)
                return
        total_pos += npos
        ua += nua
        print(
            json.dumps({"period": p, "positive_words": npos, "U_eq_A": nua}),
            flush=True,
        )
    summary = {
        "min_period": args.min_period,
        "max_period": args.max_period,
        "positive_words": total_pos,
        "U_eq_A": ua,
    }
    args.output.write_text(json.dumps(summary, indent=2) + "\n")
    print("NO_COUNTEREXAMPLE=" + json.dumps(summary, sort_keys=True), flush=True)
    print("COMPLETE: exhaustive clean range is COMPUTED, not a proof", flush=True)


if __name__ == "__main__":
    sys.exit(main())

File: experiments/issue73_20260908/counterword_search.cpp
// H-20260908-03 fast exhaustive U=A search.
// Positive-sum cyclic ±1 words; P2 via the closed-form one-period unfold.
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <string>
#include <vector>

static bool all_supplied(int p, uint32_t mask, std::vector<int> *lags_out) {
  std::vector<int> w(p);
  int S = 0;
  for (int i = 0; i < p; ++i) {
    w[i] = (mask >> i) & 1 ? 1 : -1;
    S += w[i];
  }
  if (S <= 0) return false;
  if (lags_out) lags_out->clear();
  for (int t = 0; t < p; ++t) {
    if (w[t] < 0) continue;
    std::vector<int> back(p);
    long long weight = 0;
    for (int i = 0; i < p; ++i) {
      int e = w[(t - (i + 1) % p + p) % p];
      // back[r] = sign at offset r+1
      back[i] = e;
      weight += (long long)(i + 1) * e;
    }
    long long prefix = 0, moment = 0;
    bool hit = false;
    int first_d = -1;
    for (int r = 0; r < p; ++r) {
      if (S != 0) {
        long long need = 1 - prefix;
        if (need % S == 0) {
          long long quot = need / S;
          long long d = quot * p + r;
          if (quot >= 0 && d > 0) {
            long long value = quot * weight +
                              (long long)p * S * quot * (quot - 1) / 2 +
                              quot * p * prefix + moment;
            if (value == 0) {
              hit = true;
              first_d = (int)d;
              break;
            }
          }
        }
      }
      prefix += back[r];
      moment += (long long)(r + 1) * back[r];
    }
    if (!hit) return false;
    if (lags_out) lags_out->push_back(first_d);
  }
  return true;
}

int main(int argc, char **argv) {
  if (argc != 3) {
    std::fprintf(stderr, "usage: %s min_period max_period\n", argv[0]);
    return 2;
  }
  int minp = std::atoi(argv[1]);
  int maxp = std::atoi(argv[2]);
  std::printf("protocol=H-20260908-03 exhaustive U=A C++\n");
  std::fflush(stdout);
  // control
  {
    // SSSSAAAASAAA = S=0,A=1 bits from LSB=phase0
    // word newest encoding: phase i is bit i. SSSS AAAAS AAA
    uint32_t ctrl = 0;
    const char *s = "SSSSAAAASAAA";
    int p = 12;
    for (int i = 0; i < p; ++i) if (s[i] == 'A') ctrl |= (1u << i);
    if (all_supplied(p, ctrl, nullptr)) {
      std::printf("control_failed\n");
      return 1;
    }
    std::printf("control_SSSSAAAASAAA_UA=false\n");
  }
  uint64_t total_pos = 0;
  for (int p = minp; p <= maxp; ++p) {
    if (p > 31) {
      std::fprintf(stderr, "p>31 not supported in this binary\n");
      return 2;
    }
    uint64_t npos = 0, nua = 0;
    uint64_t limit = 1ull << p;
    for (uint64_t mask = 0; mask < limit; ++mask) {
      int S = 0;
      for (int i = 0; i < p; ++i) S += ((mask >> i) & 1) ? 1 : -1;
      if (S <= 0) continue;
      ++npos;
      std::vector<int> lags;
      if (all_supplied(p, (uint32_t)mask, &lags)) {
        ++nua;
        std::string word;
        word.resize(p);
        for (int i = 0; i < p; ++i) word[i] = ((mask >> i) & 1) ? 'A' : 'S';
        std::printf("COUNTEREXAMPLE={\"period\":%d,\"word\":\"%s\",\"sign_sum\":%d,\"mask\":%llu}\n",
                    p, word.c_str(), S, (unsigned long long)mask);
        std::fflush(stdout);
        return 0;
      }
    }
    total_pos += npos;
    std::printf("{\"period\":%d,\"positive_words\":%llu,\"U_eq_A\":%llu}\n",
                p, (unsigned long long)npos, (unsigned long long)nua);
    std::fflush(stdout);
  }
  std::printf("NO_COUNTEREXAMPLE={\"min_period\":%d,\"max_period\":%d,\"positive_words\":%llu,\"U_eq_A\":0}\n",
              minp, maxp, (unsigned long long)total_pos);
  std::printf("COMPLETE: exhaustive clean range is COMPUTED, not a proof\n");
  return 0;
}
